crates/agent/src/directives/storage_mappings.rs

```diff
@@ -0,0 +1,294 @@
+    anyhow::ensure!(
+        user_has_admin,
+        "user does not have required 'admin' capability to '{}'",
+        claims.catalog_prefix
+    );
+
+    // Check that we can actually access the storage bucket before fetching (and locking) the
+    // existing `storage_mappings` rows, since this check requires multiple network round trips.
+    check_bucket_access(row.logs_token, logs_tx, &claims.add_store).await?;
+
+    let recovery_prefix = format!("recovery/{}", claims.catalog_prefix);
+    let existing_mappings =
+        fetch_storage_mappings(&claims.catalog_prefix, &recovery_prefix, txn).await?;
+    let (mut collection_data, mut recovery_data) = parse_existing(existing_mappings)?;
+
+    let mut collection_store = claims.add_store.clone();
+    // The storage mapping for collection data should always have a `collection-data/` prefix in
+    // order to segregate it from recovery log data. This makes it easier to apply bucket lifecycle
+    // policies, since you can target just the `collection-data/` prefix. If the store already
+    // specifies a prefix, then we'll add `collection-data/` to the end.
+    collection_store
+        .prefix_mut()
+        .as_mut_string()
+        .push_str("collection-data/");
+    add_store(&mut collection_data.spec, collection_store);
+    // The recovery log store doesn't need a separate prefix because all recovery log journals
+    // already begin with `recovery/`.
+    add_store(&mut recovery_data.spec, claims.add_store.clone());
+
+    Ok((collection_data, recovery_data))
+}
+
+// This is a macro instead of a function to work around the fact that file paths are `OsStr`s
+// instead of regular `&str`s.
+macro_rules! check_command {
+    ($prog:expr, $($arg:expr),* $(; region $region:expr)?) => {{
+        let mut cmd = std::process::Command::new($prog);
+        $(
+            cmd.arg($arg);
+        )*
+        $(
+          if let Some(region_name) = $region {
+              cmd.arg("--region");
+              cmd.arg(region_name);
+          }
+        )?
+        cmd
+    } };
+}
+
+async fn check_bucket_access(
+    logs_token: Uuid,
+    logs_tx: &logs::Tx,
+    store: &models::Store,
+) -> anyhow::Result<()> {
+    let mut test_file = tempfile::NamedTempFile::new().context("creating temp file")?;
+    test_file
+        .write_all(TEST_FILE_CONTENT.as_bytes())
+        .context("writing test file content")?;
+    let test_file_path = test_file.path();
+
+    let commands = match store {
+        models::Store::S3(conf) => vec![
+            (
+                "put object",
+                check_command!(
+                    "aws",
+                    "s3",
+                    "cp",
+                    test_file_path,
+                    without_query(conf.as_url()).join(TEST_FILENAME)?.to_string()
+                    ; region conf.region.as_ref()),
+            ),
+            // List comes after put, so the prefix, if configured, is guaranteed to exist.
```

Comment:
Could we add a `get` as well here? IIRC we've actually seen this misconfiguration in the past, where a user had granted list but not get.

Reply:
yep, done